src/Umbraco.Examine/ContentValueSetBuilder.cs

```diff
@@ -16,20 +20,46 @@ public class ContentValueSetBuilder : BaseValueSetBuilder<IContent>, IContentVal
     {
         private readonly UrlSegmentProviderCollection _urlSegmentProviders;
         private readonly IUserService _userService;
+        private readonly IScopeProvider _scopeProvider;
 
+        [Obsolete("Use the other ctor instead")]
         public ContentValueSetBuilder(PropertyEditorCollection propertyEditors,
             UrlSegmentProviderCollection urlSegmentProviders,
             IUserService userService,
             bool publishedValuesOnly)
+            : this(propertyEditors, urlSegmentProviders, userService, Current.ScopeProvider, publishedValuesOnly)
+        {
+        }
+
+        public ContentValueSetBuilder(PropertyEditorCollection propertyEditors,
+            UrlSegmentProviderCollection urlSegmentProviders,
+            IUserService userService,
+            IScopeProvider scopeProvider,
+            bool publishedValuesOnly)
             : base(propertyEditors, publishedValuesOnly)
         {
             _urlSegmentProviders = urlSegmentProviders;
             _userService = userService;
+            _scopeProvider = scopeProvider;
         }
 
         /// <inheritdoc />
         public override IEnumerable<ValueSet> GetValueSets(params IContent[] content)
         {
+            Dictionary<int, IProfile> creatorIds;
+            Dictionary<int, IProfile> writerIds;
+
+            // We can lookup all of the creator/writer names at once which can save some
+            // processing below instead of one by one.
+            using (var scope = _scopeProvider.CreateScope())
+            {
+                creatorIds = content.Select(x => x.CreatorId).Distinct().Select(x => _userService.GetProfileById(x))
```

Comment:
I suggest we create an extension method like `IEnumerable<IProfile> GetProfilesById(this IUserService userService, params int[] ids)`

I'm not sure how this can save processing , when it still look up one by one

Reply:
Makes sense to me 👍

Reply:
I assume it handles batches if too many ids are passed in?

Reply:
@sitereactor, the underlying `GetUsersById` handles if there are 2000+ ids, by loading all items into the cache.

Reply:
Yeah, that was a 'tanketorsk' so I deleted my comment. All good 😉 👍

Reply:
@sitereactor  👍 

Will you review the changes:
https://github.com/umbraco/Umbraco-CMS/pull/7994/commits/27f7d5efae744e9e4d227e3d526ec83113ded80f

And merge if happy

Reply:
@bergmania 

> I'm not sure how the current implementation can save processing , when it still look up one by one.

It's because this is doing a `Distinct` so "in theory" it could reduce processing. All users are cached. I don't think `GetUsersById` goes into cache since the repository cache policy is just the default so this will cause even more queries to execute

Reply:
All users are cached only when they are resolved by single ID based on the UserRepository cache policy